Dolphin new Dump SSL features

sepalani opened the pull request:
This pull request adds the following SSL features and are modifiable through the Dolphin INI configuration file:
 - **SSLDumpRead:** Dump ssl_read content into the file GAMEID_read.bin
 - **SSLDumpWrite:** Dump ssl_write content into the file GAMEID_write.bin
 - **SSLVerifyCert:** This allows Dolphin to verify the SSL certificate (like the Wii)
 - **SSLDumpRootCA:** Dump the root CA into the file HOSTNAME_rootca.der
 - **SSLDumpPeerCert:** Dump the peer cert into the file HOSTNAME_peercert.der

Those are all disabled by default and won't mess current games' settings. A new folder _("SSL")_ has been added into the Dump directory to store them.

**EDIT:**
 - ERROR_LOG: SSL handshake error's details
 - A security profile to allow low RSA key was added (this issue came when upgrading mbedtls)
 - Missing init and free call added

Ready to be reviewed & merged.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/2868)
<!-- Reviewable:end -->

phire commented:
Something has confused buildbot. Looks like it's failing to merge.

BhaaLseN commented:
Can you please rebase on top of upstream/master, instead of merging it in?

phire commented:
Looks like your last commit is at fault. Could you rebase on the latest master and delete the merge of master into your branch.

lioncash commented on Source/Core/Common/CommonPaths.h:
```diff
@@ -68,6 +68,7 @@
 #define DUMP_FRAMES_DIR     "Frames"
 #define DUMP_AUDIO_DIR      "Audio"
 #define DUMP_DSP_DIR        "DSP"
+#define DUMP_SSL_DIR		"SSL"
```
tab space mismatch

lioncash commented on Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_net_ssl.cpp:
```diff
@@ -255,6 +259,12 @@ IPCCommandResult CWII_IPC_HLE_Device_net_ssl::IOCtlV(u32 _CommandAddress)
+			if (SConfig::GetInstance().m_SSLDumpRootCA)
+			{
+				std::string fname = File::GetUserPath(D_DUMPSSL_IDX) + ssl->hostname + "_rootca.der";
+				File::IOFile(fname.c_str(), "wb").WriteBytes(Memory::GetPointer(BufferOut2), BufferOutSize2);
```
calling c_str is unnecessary here.

sepalani commented:
Done. It's rebased, and I fixed the tab alignment and the useless c_str conversion @lioncash mentioned.

delroth commented:
We have a PCAP writer class in Common/. Please dump the data into PCAPs instead of a custom format.

randomstuff commented:
@delroth, At this layer, we don't have any notions of packet. Is there any suitable [Link-Layer type](http://www.tcpdump.org/linktypes.html)? The data would have to be re-encapsulated in fake TCP frames (or something similar).

delroth commented:
I would just use NULL. PCAPs are convenient because you get the messages interlaced and the convenience of pre-existing tools.

randomstuff commented:
@delroth, If you use NULL you're supposed to prepent each packet by a PF_* value (such as PF_INET if encapsulated in IPv4 or PF_INET6 for IPv6…).

Tilka commented:
Any encapsulation effort is worth the payoff of getting pcaps, imho. If you want to get the commits unrelated to dumping merged earlier, feel free to open a separate PR.

vgturtle127 commented:
So, if I'm not mistaken, this will allow us to dump SSL certificates used by games? How will that work with most of the games being offline, etc.? I figured they were not stored on the disc, and they were retrieved when needed or something? Unless you can still download the certificates from the servers with NUS downloader or something. Or do I not understand how this works at all?

Also, what versions of SSL/TLS are supported (on the Wii and in this PR), and could we use a spoofed certificate and use SSL_SETCLIENTCERTDEFAULT to emulate all of the known certificates? I'm not sure if all the certs have been dumped or not. Unless I am totally misunderstanding what this does. I can only assume. :blush:

sepalani commented:
This will help to dump SSL traffics and certificates. Before NintendoWFC shutdown, a special version of Dolphin was used to dump SSL traffic without using a SSL Tunnel or a MITM. It was just logging ssl_read and ssl_write content.

The goal of this PR is to add SSL features into Dolphin such as dumping the peer certificate (i.e. the server one), the rootCA certificate (i.e. the in-game one, usually stored into the main.dol), force Dolphin to verify the SSL certificates (which is the Wii's default behavior) and of course logging ssl_read and ssl_write. I added into the INI config file a way to enable each one of them independently. FTM, I didn't plan to add tricky things like spoofing the SSL certificate or force a specific SSL version and so on.

Why someone would like to only dump ssl_write? Because this person may own the server. Then, why someone would like to only dump ssl_read? Someone may run an homebrew and only want to know what the server is sending without having to print tons of debug. If you want to replace an in-game certificate and check if it's working on real a Wii (which verifies SSL certificates) using Dolphin, it has to verify the certificate and returns the proper error code. In sum, that's the purpose of this PR.

Otherwise, dumping them in a PCAP file is a nice idea, however that wasn't my main concern. Honestly, even if I have to make a PCAP file for that, I will probably do this in another PR and add a way to log ICMP, DNS with it to simulate as much as possible a real network dump.

Concerning the games, well... This PR was more for developers than lambda users. It might be useful to reverse games that use SSL servers and which aren't implemented yet like Monster Hunter 3 for example that uses Capcom servers or some CoD games that use Daemonware servers.

vgturtle127 commented:
@sepalani Oh, alright. Well, not many games supported online multiplayer right? Is there a list of games that do so we can try them all?

Thanks for all the information, by the way. :smile:

sepalani commented:
You can try it on all online games (this list isn't exhaustive): http://www.nintendo.com/consumer/gameslist/en/games-wifi.jsp 

A bit useless for most of them because Wiimmfi and Altwfc servers use SSL Stripping to patch games (and their SSL connections to Nintendo servers). Then you are able to play online via alternative servers.

If you want to patch the SSL certificate of an existing game you'll have to find it in the game, extract it and replace it with your own carefully crafted. Usually it's in DER format and you can track it while searching in binary mode into the game filesystem the RootCA certificate this PR's Dolphin dumps . The only game where I tested this feature was Monster Hunter Tri.

vgturtle127 commented:
Thanks, I am going to use this PR and the info you gave me to do some research on the Nintendo WiFi connection and see if I can get some emulation going. I think NO$GBA, Desmume, and Dolphin have most of the NFC stuff covered, but private alternate servers are something that haven't been completely explored. Anyway, thanks for the info. :smile:

sepalani commented:
PR rebased according to the new mbedtls. Still ready to be reviewed & merged.

delroth commented:
@dolphin-emu-bot rebuild

sepalani commented:
I added 2 more commits.

The purpose of the first one is to add more clarity in debug/error logs via the proper mbedtls error functions.

The second one fixes an issue that comes from the migration from polarssl to mbedtls. With mbedtls we can define a security profile and by default it doesn't accept RSA keys lower than 2048-bit, even if it's correct. I quickly checked on my Wii and as I thought the Wii doesn't check this at all. The main issue is in most network dumps I have, I see 1024-bit RSA keys and sometimes 2048-bit ones. This is why I get rid of this check, that the Wii doesn't have anyway.

I also used mbedtls_ssl_get_peer_cert instead of grabbing the cert on the fly from the struct. I tried to replace the code I made before with the appropriate mbedtls functions when they exist. Of course, I rebased and tested this PR with the latest version of Dolphin.

Ready to reviewed and merged.

degasus commented:
PCAP might be nicer, but the dumping way here doesn't look as bad. Especially as it's an update for the already merged DUMP_SSL makro. SO for me, this PR looks good.

@delroth What is your opinion, shall we just merge it as it is?

sepalani commented on Source/Core/Core/IPC_HLE/WII_Socket.cpp:
```diff
@@ -313,8 +315,14 @@ void WiiSocket::Update(bool read, bool write, bool except)
 					{
-
-						int ret = mbedtls_ssl_handshake(&CWII_IPC_HLE_Device_net_ssl::_SSL[sslID].ctx);
+						mbedtls_ssl_context *ctx = &CWII_IPC_HLE_Device_net_ssl::_SSL[sslID].ctx;
```
Pointer against the type name.

sepalani commented:
The typo is fixed and this PR is rebased against latest version of master.

However, as always it keeps breaking SSL features. With the latest version of master, the **IOCTLV_NET_SSL_DOHANDSHAKE** is lost. It's still working with [Dolphin 4.0-7966](https://github.com/sepalani/dolphin/tree/stable_ssl).

I'm going to investigate what is breaking the SSL handshake. You can merge this PR but you won't be able to test the ssl_read/write features because ssl communication is broken due to this abnormal handshake behavior.

It doesn't seem to be my SSL server fault, old versions of Dolphin can still connect, ditto for old versions of this PR using ```git reflog``` to fall back on a working version.

bb010g commented:
bump

vgturtle127 commented:
:+1:

mathieui commented:
@dolphin-emu-bot rebuild

sepalani commented:
Rebase against latest master.

I had lots of diff with tab alignment replaced with spaces. I assumed the coding style has changed and adapted the code consequently, if it's not the case I can rollback to the previous coding style using ```git reflog```.

Meanwhile, I discovered there were missing init and free calls so I added them. I also changed the order to match the one used in mbedtls examples (ssl_client2). I don't think that's important but just in case.

Otherwise, still ready to be reviewed and merged.

JosJuice commented:
Yes, the coding style has changed. Now, everything that isn't formatted according to clang-format is rejected by lint.

sepalani commented:
I installed clang-format plugin for Visual Studio, the coding style issues should be fixed now.

JosJuice commented:
@dolphin-emu-bot rebuild

sepalani commented:
I forgot a free at [line 176](https://github.com/dolphin-emu/dolphin/pull/2868/files#diff-a1fa55b1731a901eb63a7ba241cea1ebR176), then I figured out that [this string](https://github.com/dolphin-emu/dolphin/pull/2868/files#diff-a1fa55b1731a901eb63a7ba241cea1ebR169) can be static const. I didn't find any constexpr strlen function, but it shouldn't be too big of deal since many compilers replace strlen builtin with static const strings length directly.

Fallcrest commented:
Could you rebase and make sure it works with the new dump path tool ([3944](https://github.com/dolphin-emu/dolphin/pull/3944))? I am sure it does, but it needs to be rebased anyhow.

sepalani commented:
@Fallcrest 
Done.

RisingFog commented:
@dolphin-emu-bot rebuild

Fallcrest commented:
With all of the new debug features being worked on, I was wondering if this could get a final once over and possibly get merged. It would be a nice tool to have.

shuffle2 commented:
pcap would be nice, but as @degasus said. merge anyway.

vgturtle127 commented:
@sepalani Is there a repository of certificates or a Wiki with certificate info or something? I would be curious to research this...

sepalani commented:
To be honest I don't know. I'm sure with a google hack you'll be able to find the answer pretty quickly. However, if you're talking about private certificates, I'm sure you won't find anything.

vgturtle127 commented:
@sepalani Alright, I will try to look a little deeper and see what I can find. Thanks!

😄
